core/finality-grandpa/src/environment.rs

```diff
@@ -529,9 +534,11 @@ where
 	}
 
 	fn proposed(&self, _round: u64, propose: PrimaryPropose<Block>) -> Result<(), Self::Error> {
-		let local_id = self.config.local_key.as_ref()
+		let local_id = self.session_store.keys()
```

Comment:
I don't think this or the logic in`prevoted`, etc. is right.

`round_data` chooses which key to sign messages with (i.e. which key is _actually_ voting). But this may return a completely different identity in the case that the session store's underlying key set has changed in the meantime. And then we'd store the vote to disk as if _that_ identity signed the vote.

We should probably select a key upon construction of an `Environment`. This is done every time the validator set changes anyway.